### ISSUE TITLE
 [SMARTSCAN] [$500] LOW: Android - Scan-When adding big image, thumbnail display is not clear, image distorted

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#32649]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.9-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:** @

## Action Performed:
1. Launch app
2. Tap fab---Request money--Scan
3. Upload below given image
4. Select a user
5. Tap request
6. Tap request created
7. Note the thumbnail receipt

## Expected Result:
When adding big image, thumbnail display must be clear.

## Actual Result:
When adding big image as scan receipt, thumbnail display is not clear, image distorted.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

![Bug6304241_1701940730946!Screenshot_2023-12-07-09-49-02-611_com expensify chat](https://github.com/Expensify/App/assets/115492554/93355116-330d-4ee8-ba93-b9b4cd9c1de5)

![Bug6304241_1701940730965!bigggg-2023-11-23_19_56_43 485](https://github.com/Expensify/App/assets/115492554/37110fd0-0d69-4e91-bdee-6ce044fb1529)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0122f62b0e85d3a0e6</li>
        <li>Upwork Job ID: 1732723141179482112</li>
        <li>Last Price Increase: 2023-12-07</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Julesssss</details>

**Comment by ZhenjaHorbach** [comment]
# Proposal
## Please re-state the problem that we are trying to solve in this issue.

Scan-When adding big image, thumbnail display is not clear, image distorted

## What is the root cause of that problem?

<img width="394" alt="Screenshot 2023-12-07 at 14 11 11" src="https://github.com/Expensify/App/assets/68128028/d374b517-4cfc-4334-bb79-edd1f2e74a44">
This is actually expected behavior
In addition to the original photos
We use thumbnail versions
The height of which is (1024)
And if we have such a long photo
We get such funny values (37x1024)


## What changes do you think we should make in order to solve the problem?

The most important thing is we need to have access to better images
Which are stored on the backend

But from the frontend side we can use the original image in case if, for example, the photo width is more than 500 pixels

For this we can add new condition and use `Image `

```

const [isBlurredThumbnail,setBlurredThumbnail] = useState(false)

	Image.getSize(thumbnailSource, (width, height) => {
		if(width < 500) {
		    setBlurredThumbnail(true)
		}
	});
	
	
```

https://github.com/Expensify/App/blob/10727a7979709987c77e94a2d059f56bd09fa5eb/src/components/ReportActionItem/ReportActionItemImage.js#L49-L73

And then if isBlurredThubbnail is true we will use imageSource instead thumbnailSource

https://github.com/Expensify/App/blob/10727a7979709987c77e94a2d059f56bd09fa5eb/src/components/ReportActionItem/ReportActionItemImage.js#L67

## What alternative solutions did you explore? (Optional)
NA

**Comment by abekkala** [comment]
@yuwenmemon just wanted to get your opinion on this
https://github.com/Expensify/App/issues/32649#issuecomment-1852200023

**Comment by yuwenmemon** [comment]
Yeah I'd rather go with cropping the image for a thumbnail on the backend than the proposed solution. Let's make this internal.

**Comment by abekkala** [comment]
tagged Greg in #wave6-collect-submitters to get confirmation to add this to that project

**Comment by abekkala** [comment]
added to wave 5 project

**Comment by abekkala** [comment]
@mallenexpensify 
### STATUS: 
I'm ooo through Jan 03

@dylanexpensify  did a call-out for an internal engineer to take this on.

**Comment by mallenexpensify** [comment]
hahahah, I'm out next week too.  I'll either check in once or twice for my issues or reassign.

**Comment by mallenexpensify** [comment]
Unassigning cuz also out, we messaged folks in #expensify-open-source to let them know most BZ would be off for the week.

**Comment by abekkala** [comment]
@dylanexpensify do we have an internal engineer willing to take this on for your project

**Comment by abekkala** [comment]
waiting for internal engineer

**Comment by Julesssss** [comment]
We can no longer download the large image attached above... Github shows the image as not available when I try to open it or download it 😕 

I found two alternative images: 
- With image 1 below I'm not seeing any image quality reductions, but this is a much shorter image
- With image 2 the app is crashing, though that is a different bug from the original report 

**Testing image 1**

https://github.com/Expensify/App/assets/10736861/47fb47f0-67c2-4202-a268-56147dbbec63

**Image 1**
![large-image-2](https://github.com/Expensify/App/assets/10736861/21b0d8ea-1589-4f81-b088-758d620d593d)

**Testing image 2**

https://github.com/Expensify/App/assets/10736861/a4553c10-de49-4caa-89fb-a307a28ad4d0


**image 2**
![large-image-1](https://github.com/Expensify/App/assets/10736861/18b6cbca-c899-4e66-a475-052ca413ba75)

**Comment by Julesssss** [comment]
Here's the Android Exception: we previously resolved a similar issue by switching image libraires and I'll probably handle this separately. 

```
FATAL EXCEPTION: main
    Process: com.expensify.chat, PID: 28882
    java.lang.RuntimeException: Canvas: trying to draw too large(149730560bytes) bitmap.
    	at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:266)
    	at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:94)
    	at android.graphics.drawable.BitmapDrawable.draw(BitmapDrawable.java:549)
    	at android.widget.ImageView.onDraw(ImageView.java:1446)
    	at expo.modules.image.ExpoImageView.onDraw(ExpoImageView.kt:204)
    	at android.view.View.draw(View.java:23900)
    	at expo.modules.image.ExpoImageView.draw(ExpoImageView.kt:200)
    	at android.view.View.updateDisplayListIfDirty(View.java:22767)
    	at android.view.View.draw(View.java:23631)
    	at android.view.ViewGroup.drawChild(ViewGroup.java:4559)
    	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4320)
    	at android.view.View.updateDisplayListIfDirty(View.java:22758)
    	at android.view.View.draw(View.java:23631)
    	at android.view.ViewGroup.drawChild(ViewGroup.java:4559)
    	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4320)
```

**Comment by Julesssss** [comment]
I'll probably work on this in parallel with [this](https://github.com/Expensify/App/issues/34372) related crash.

**Comment by Julesssss** [comment]
Beneath a height/width ratio of about 1:3, the thumbnail becomes indistinguishable:

Image1: `1904/19660` > `99/1024` ❌
Image2: `730/4283` > `175/1024` ❌
Image3:  `730/2603` > `287/1024` ❌ 
Image4: `730/1796` > `416/1024` ✅ 

We currently lock thumbnail resolution to 1024px max. To resolve the issue we'll need to apply further logic to images with a w/h ratio greater than 3:1 in either width or height. I see a few options:

1) Crop the image at the center -- this keeps thumbnails small, but the preview doesn't show the true height/width ✅ 
2) Increase the max width from 1024 -- this isn't ideal as very tall or wide images will have the same problem with low-resolution thumbnails ❌ 
3) Prevent users from uploading images above a certain resolution -- bad as users  receipts will fail and need to be edited manually ❌
4) Discard the thumbnail and display the full-resolution image in NewDot -- workaround ❌

_NewDot thumbnails are displayed much larger than OldDot, so we should probably limit these changes to NewDot to avoid breaking OldDot_ 

note: our thumbnail suffix for avatars is `_128.jpeg`

**Comment by Julesssss** [comment]
When is thumbnail created?

A) OldDot Receipt: Scaled to fit into rectangular receipt
B) OldDot Attachment: Displayed unedited 
C) NewDot Receipt: cropped & blurry
D) NewDot Receipt preview: cropped
E) NewDot Attachment: Displayed unedited 

/ | Image
-- | --
A | ![Screenshot 2024-01-18 at 15 20 26](https://github.com/Expensify/App/assets/10736861/4021a0e3-56ac-4a11-ac94-80dd475b4f62)
B | ![Screenshot 2024-01-18 at 15 20 31](https://github.com/Expensify/App/assets/10736861/16d4ea13-f5dd-4792-8f8d-b622b4ceaf1e)
C | ![Screenshot 2024-01-18 at 15 26 43](https://github.com/Expensify/App/assets/10736861/069c5f73-f154-460f-b498-db63a3b089fd)
D | ![Screenshot 2024-01-18 at 15 25 27](https://github.com/Expensify/App/assets/10736861/3129d5f3-c141-4336-9ad6-d9632340b4d1)
E | ![Screenshot 2024-01-18 at 15 24 39](https://github.com/Expensify/App/assets/10736861/fc92dd83-6f3b-486c-a2d6-077bc0d5ef84)

**Comment by Julesssss** [comment]
Based on the above, the best course of action is to crop image previews, providing they are NewDot receipts. We still want to keep the scale for tall/wide image attachments.  

In both OldDot and NewDot, receipts AND image attachments pass through the [same Web-PDF code](https://github.com/Expensify/Web-PDFs/blob/58307e078a49eb43a60dddcece4d7eab768a959c/var/pdfs.expensify.com/lib/uploadReceipt.php#L375-L386). The next step is to figure out how to crop with [Imagemagick CLI tools ](https://imagemagick.org/script/command-line-processing.php)and how to pass a `shouldCrop` flag. 

**Current Resize logic:** 
```
/usr/bin/convert -background white -flatten -limit memory 128mb -limit map 128mb -scale '{$geometry}x{$geometry}> [inputfile] [outputfile]
```

**Comment by Julesssss** [comment]
I made more progress on [the PR](https://github.com/Expensify/Web-PDFs/pull/560) today: 
- Logic is in place for handling images both larger and smaller than the desired width/height
- I debugged all the NewDot/OldDot attachment/Request flows in both Web-E & Web-PDFs
- A shouldCropReceipt param from WEB is being passed down to the Web-PDF `createReceipt` function  

**Next Steps:**
- Set the `createReceipt` param in Web-E dynamically
- Check whether ImageMagic scale function can continue to be used for non-cropped images
- Thoroughly test varients of image types and pdfs
- Open up PRs

**Comment by Julesssss** [comment]
Implementation is complete: 

- The [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/40622) that introduces the request param is currently in review
- The main [Web-PDFs PR](https://github.com/Expensify/Web-PDFs/pull/560) is on hold, but the code and tests are completed

_There is a new CI issue within Web-PDFs that I need to fix._

**Comment by Julesssss** [comment]
I'm ill today, but updated the Web-E PR based on feedback. It should be merged today 🤞

**Comment by Julesssss** [comment]
I've been sick. I'm back 50% now and trying to get the 2nd half of this task merged.

**Comment by Julesssss** [comment]
I resolved the CI issue and the [full PR](https://github.com/Expensify/Web-PDFs/pull/560) is now in review 🎉

**Comment by Julesssss** [comment]
[The PR](https://github.com/Expensify/Web-PDFs/pull/560) received a small change request which I have made, it's now back in review.
